### PR TITLE
Infer output shape of Expand op

### PR DESCRIPTION
**Pull request by shinh**

### CODE REVIEW

**Comment by shinh** [comment]
@houseroad : Can I request you to review this PR since you have reviewed a few PRs I made for shape inference?

**Comment by houseroad** [comment]
Also could you please add a test case to https://github.com/onnx/onnx/blob/master/onnx/test/shape_inference_test.py

**Comment by shinh** [comment]
Ah, I thought this op is `np.broadcast_to`. Now it handles mutlidirectional broadcasting. I've added tests for them, too. Thanks for your review!

**Comment by shinh** [comment]
Ping. Could you take another look?

File: onnx/defs/math/defs.cc [hunk 1]
```diff
@@ -1085,7 +1085,50 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+            "Constrain input and output types to all tensors.")
+        .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
+          // Type inference
+          propagateElemTypeFromInputToOutput(ctx, 0, 0);
+          // Shape Inference if 2nd input data (the target shape) is available
+          const TensorProto* target_shape_initializer = ctx.getInputData(1);
+          if (!target_shape_initializer) {
+            return;
+          }
+          // The targetShape vector represents the specified shape for output.
```
Comment (houseroad) [inline onnx/defs/math/defs.cc]: nit: targetShape => target_shape

File: onnx/defs/math/defs.cc [hunk 2]
```diff
@@ -1085,7 +1085,50 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+          } else {
+            const auto& data = target_shape_initializer->int64_data();
+            target_shape.insert(target_shape.end(), data.begin(), data.end());
+          }
+          auto& input_shape = getInputShape(ctx, 0);
+          auto* output_shape =
+              ctx.getOutputType(0)->mutable_tensor_type()->mutable_shape();
+          for (int i = 0; i < input_shape.dim_size(); ++i) {
+            auto* new_dim = output_shape->add_dim();
+            int64_t target_dim = static_cast<size_t>(i) < target_shape.size() ? target_shape[i] : 1;
```
Comment (houseroad) [inline onnx/defs/math/defs.cc]: Dimensions are right aligned. Please check the case here: https://github.com/onnx/onnx/blob/master/docs/Operators.md#expand

File: onnx/test/shape_inference_test.py [hunk 3]
```diff
@@ -234,6 +234,33 @@ def test_reshape_static_shape_constant(self):  # type: () -> None
...
+        self._assert_inferred(graph, [make_tensor_value_info('y', TensorProto.UINT8, (2, 4, 3))])
+
+    def test_expand_static_shape_inferred_broadcast(self):  # type: () -> None
+        graph = self._make_graph(
+            [('x', TensorProto.UINT8, (2, 1, 3)),
+             ('shape', TensorProto.INT64, (3,))],
+            [make_node("Expand", ['x', 'shape'], ['y'])],
+            [],
+            initializer=[make_tensor('shape', TensorProto.INT64, (2,), (1, 4))])
+        self._assert_inferred(graph, [make_tensor_value_info('y', TensorProto.UINT8, (2, 4, 3))])
```
Comment (houseroad) [inline onnx/test/shape_inference_test.py]: this is an incorrect case.

File: onnx/test/shape_inference_test.py [hunk 4]
```diff
@@ -234,6 +234,33 @@ def test_reshape_static_shape_constant(self):  # type: () -> None
...
+
+    def test_expand_static_shape_inferred_incomplete(self):  # type: () -> None
+        graph = self._make_graph(
+            [('x', TensorProto.UINT8, ('d1', 'd2', 'd3')),
+             ('shape', TensorProto.INT64, (3,))],
+            [make_node("Expand", ['x', 'shape'], ['y'])],
+            [],
+            initializer=[make_tensor('shape', TensorProto.INT64, (2,), (2, 1))])
+        self._assert_inferred(graph, [make_tensor_value_info('y', TensorProto.UINT8, (2, 'd2', 'd3'))])
```
Comment (houseroad) [inline onnx/test/shape_inference_test.py]: this should be d1, 2, d3

File: onnx/defs/math/defs.cc (hunk 1 above)
Comment (shinh) [reply onnx/defs/math/defs.cc]: Done

File: onnx/defs/math/defs.cc (hunk 2 above)
Comment (shinh) [reply onnx/defs/math/defs.cc]: Oops, sorry for the incorrect understanding. Done

File: onnx/test/shape_inference_test.py (hunk 4 above)
Comment (shinh) [reply onnx/test/shape_inference_test.py]: Done

File: onnx/test/shape_inference_test.py (hunk 3 above)
Comment (shinh) [reply onnx/test/shape_inference_test.py]: Done

**Comment by shinh** [comment]
> please add a test in which shape's dim is larger than input's dim.

Sorry, I missed this comment. Added a test and fixed the implementation. I hope I'm not missing anything now..

**Comment by shinh** [comment]
Ping? Are there still issues?

**Comment by shinh** [comment]
@houseroad friendly reminder! Is there still any issue?

**Comment by shinh** [comment]
It seems this was fixed by https://github.com/onnx/onnx/pull/2041

**Pull request closed without merging**